synapse/logging/context.py

```diff
@@ -624,6 +626,11 @@ def make_deferred_yieldable(deferred):
 
     (This is more-or-less the opposite operation to run_in_background.)
     """
+    if inspect.isawaitable(deferred):
+        # If we're given a coroutine we need to convert it to a deferred so that
+        # we can attach callbacks (and not immediately return).
```

Comment:
I'm not convinced that's the reason we need to convert it to a deferred (we can do similar things by awaiting awaitables).

What it does do is to start running the coroutine and find out if it has finished or not, which saves us clearing and immediately restoring the logcontext in the common case that the coroutine completes synchronously.

Reply:
Hmm, fair. I guess I was thinking of this in terms of converting to a future/deferred so that we could put the callbacks on it (since you can't do that on raw coroutines). Will reword